sno/diff.py

```diff
@@ -428,97 +427,90 @@ def diff(ctx, output_format, output_path, exit_code, args):
+            target_rs = RepositoryStructure.lookup(repo, commit_parts[2] or "HEAD")
+            working_copy = None
+        else:
+            # When one commit is specified, it is base, and we diff base<>working_copy.
+            # When no commits are specified, base is HEAD, and we do the same.
+            # We diff base<>working_copy by diffing base<>target + target<>working_copy,
+            # and target is set to HEAD.
+            base_rs = RepositoryStructure.lookup(repo, commit_parts[0])
+            target_rs = RepositoryStructure.lookup(repo, "HEAD")
             working_copy = WorkingCopy.open(repo)
             if not working_copy:
                 raise NotFound(
                     "No working copy, use 'checkout'", exit_code=NO_WORKING_COPY
                 )
+            working_copy.assert_db_tree_match(target_rs.tree)
 
-            working_copy.assert_db_tree_match(commit_head.peel(pygit2.Tree))
+        # Parse [<dataset>[:pk]...]
+        paths = {}
+        for p in args:
+            pp = p.split(":", maxsplit=1)
+            paths.setdefault(pp[0], [])
+            if len(pp) > 1:
+                paths[pp[0]].append(pp[1])
 
-        L.debug("commit_base=%s", commit_base.id)
+        base_str = base_rs.id
+        target_str = "working-copy" if working_copy else target_rs.id
+        L.debug('base=%s target=%s', base_str, target_str)
 
         # check whether we need to do a 3-way merge
-        c_target = commit_target or commit_head
-        merge_base = repo.merge_base(commit_base.oid, c_target.oid)
-        L.debug("Found merge base: %s", merge_base)
-
-        if not merge_base:
-            # there is no relation between the commits
-            raise InvalidOperation(
-                f"Commits {commit_base.id} and {c_target.id} aren't related."
-            )
-        elif merge_base not in (commit_base.id, c_target.id):
-            # this needs a 3-way diff and we don't support them yet
-            raise NotYetImplemented(f"Sorry, 3-way diffs aren't supported yet.")
+        if base_rs.head_commit and target_rs.head_commit:
+            merge_base_id = repo.merge_base(base_rs.id, target_rs.id)
+            L.debug("Found merge base: %s", merge_base_id)
+
+            if not merge_base_id:
+                # there is no relation between the commits
+                raise InvalidOperation(
+                    f"Commits {base_rs.id} and {target_rs.id} aren't related."
+                )
+            elif merge_base_id not in (base_rs.id, target_rs.id):
+                # this needs a 3-way diff and we don't support them yet
+                raise NotYetImplemented(f"Sorry, 3-way diffs aren't supported yet.")
 
-        base_rs = RepositoryStructure(repo, commit_base)
         all_datasets = {ds.path for ds in base_rs}
-
-        target_rs = RepositoryStructure(repo, commit_target or commit_head)
-        all_datasets |= {ds.path for ds in target_rs}
+        if target_rs:
```

Comment:
AFAICT this is always truthy, it's set to a RespositoryStructure instance in both branches above, and `bool(rs)` is always True

Reply:
Done